demos/shadow/shadow_demo_main/shadow_demo_helpers.c

```diff
@@ -0,0 +1,872 @@
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file shadow_demo_helpers.c
+ * @brief This file provides helper functions used by shadow demo application to
+ * do MQTT operation based on mutually authenticated TLS connection.
+ *
+ * A mutually authenticated TLS connection is used to connect to the AWS IoT
+ * MQTT message broker in this example. Define ROOT_CA_CERT_PATH,
+ * CLIENT_CERT_PATH, and CLIENT_PRIVATE_KEY_PATH in demo_config.h to achieve
+ * mutual authentication.
+ */
+
+/* Standard includes. */
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* shadow includes */
+#include "shadow_demo_helpers.h"
+
+/* POSIX includes. */
+#include <unistd.h>
+
+/* Include Demo Config as the first non-system header. */
+#include "demo_config.h"
+
+/* OpenSSL sockets transport implementation. */
+#include "openssl_posix.h"
+
+/* Reconnect parameters. */
+#include "transport_reconnect.h"
+
+/* Clock for timer. */
+#include "clock.h"
+
+
+/**
+ * These configuration settings are required to run the mutual auth demo.
+ * Throw compilation error if the below configs are not defined.
+ */
+#ifndef AWS_IOT_ENDPOINT
+    #error "Please define AWS IoT MQTT broker endpoint(AWS_IOT_ENDPOINT) in demo_config.h."
+#endif
+#ifndef ROOT_CA_CERT_PATH
+    #error "Please define path to Root CA certificate of the MQTT broker(ROOT_CA_CERT_PATH) in demo_config.h."
+#endif
+#ifndef CLIENT_CERT_PATH
+    #error "Please define path to client certificate(CLIENT_CERT_PATH) in demo_config.h."
+#endif
+#ifndef CLIENT_PRIVATE_KEY_PATH
+    #error "Please define path to client private key(CLIENT_PRIVATE_KEY_PATH) in demo_config.h."
+#endif
+#ifndef CLIENT_IDENTIFIER
+    #error "Please define a unique CLIENT_IDENTIFIER."
+#endif
+
+/**
+ * Provide default values for undefined configuration settings.
+ */
+#ifndef AWS_MQTT_PORT
+    #define AWS_MQTT_PORT    ( 8883 )
+#endif
+
+#ifndef NETWORK_BUFFER_SIZE
+    #define NETWORK_BUFFER_SIZE    ( 1024U )
+#endif
+
+/**
+ * @brief Length of MQTT server host name.
+ */
+#define AWS_IOT_ENDPOINT_LENGTH             ( ( uint16_t ) ( sizeof( AWS_IOT_ENDPOINT ) - 1 ) )
+
+/**
+ * @brief Length of client identifier.
+ */
+#define CLIENT_IDENTIFIER_LENGTH            ( ( uint16_t ) ( sizeof( CLIENT_IDENTIFIER ) - 1 ) )
+
+/**
+ * @brief ALPN protocol name for AWS IoT MQTT.
+ *
+ * This will be used if the AWS_MQTT_PORT is configured as 443 for AWS IoT MQTT broker.
+ * Please see more details about the ALPN protocol for AWS IoT MQTT endpoint
+ * in the link below.
+ * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
+ */
+#define ALPN_PROTOCOL_NAME                  "\x0ex-amzn-mqtt-ca"
```

Comment:
Why is the first char \x0e ?

Reply:
For the shadow demo, I didn't really notice this configuration because it's a user of mqtt protocol. I will ask the writer of this configuration in mqtt demo.

Reply:
After consulting with the original writer, George Thomas, it is an open ssl requirement for ALPN string to be in wire format.
\x0e is the length of the ALPN string, the length of "x-amzn-mqtt-ca" is 14, and hex value is added to it in the beginning, which is 0e.
He will add few comment to explain that.